### PR TITLE
Feat: Add methods to project multiple points

### PR DESCRIPTION
**Pull request by EdYazbec**
New Line functions:
   - Line.project_points: Project multiple points onto the same line
   - Line.distance_points: Get the distance from multiple points to the same line

New Plane Functions:
   - Plane.project_points: Project multiple points onto the same plane
   - Plane.distance_points: Get the distance from multiple points to a plane
   - Plane.distance_points_signed: Get the signed distance from multiple points to a plane

This is functionality that I use a lot. My tests check that the new functions return the same result as the existing functions.

### CODE REVIEW

**Review by CristianoPizzamiglio — APPROVED** [review]
I think these are very useful methods. As far as I am concerned, I often have to project `Points` on a plane so I will definitely use the `project_points` method.
The code seems to be very clean to me and in line with the codebase's coding style.

File: README.rst [hunk 1]
```diff
@@ -28,6 +28,22 @@
...
+New Line functions:
+   - Line.project_points: Project multiple points onto the same line
+   - Line.distance_points: Get the distance from multiple points to the same line
+
+New Plane Functions:
+   - Plane.project_points: Project multiple points onto the same plane
+   - Plane.distance_points: Get the distance from multiple points to a plane
+   - Plane.distance_points_signed: Get the signed distance from multiple points to a plane
+
+
```
Comment (ajhynes7) [inline README.rst:46]: This shouldn't be added to the README. If you're explaining the purpose of this branch, it should go in the PR description instead.

File: src/skspatial/objects/line.py [hunk 2]
```diff
@@ -302,6 +302,46 @@ def project_point(self, point: array_like) -> Point:
         # Add the projected vector to the point on the line.
         return self.point + vector_projected
 
+    def project_points(self, points: array_like) -> Points:
+        """
+        Project points onto the line.
```
Comment (ajhynes7) [inline src/skspatial/objects/line.py:307]:
```suggestion
        Project multiple points onto the line.
```

File: src/skspatial/objects/line.py [hunk 3]
```diff
@@ -409,6 +449,43 @@ def distance_point(self, point: array_like) -> np.float64:
 
         return point_projected.distance_point(point)
 
+    def distance_points(self, points: array_like) -> np.ndarray:
+        """
+        Return the distance from points to the line.
```
Comment (ajhynes7) [inline src/skspatial/objects/line.py:454]:
```suggestion
        Return the distances from multiple points to the line.
```

File: src/skspatial/objects/line.py [hunk 4]
```diff
@@ -409,6 +449,43 @@ def distance_point(self, point: array_like) -> np.float64:
...
+
+        This is the distance from points to their projection on the line.
```
Comment (ajhynes7) [inline src/skspatial/objects/line.py:456]:
```suggestion
        These are the distances from the points to their projections on the line.
```

File: src/skspatial/objects/line.py [hunk 5]
```diff
@@ -409,6 +449,43 @@ def distance_point(self, point: array_like) -> np.float64:
...
+
+        Parameters
+        ----------
+        points : array_like
+            Input points.
+
+        Returns
+        -------
+        np.ndarray
+            Distances from the point to the line.
```
Comment (ajhynes7) [inline src/skspatial/objects/line.py:466]:
```suggestion
            Distances from the points to the line.
```

File: src/skspatial/objects/plane.py [hunk 6]
```diff
@@ -273,6 +273,37 @@ def project_point(self, point: array_like) -> Point:
 
         return Point(point) + vector_projected
 
+    def project_points(self, points: array_like) -> Points:
+        """
+        Project points onto the plane.
```
Comment (ajhynes7) [inline src/skspatial/objects/plane.py:278]:
```suggestion
        Project multiple points onto the plane.
```

File: src/skspatial/objects/plane.py [hunk 7]
```diff
@@ -273,6 +273,37 @@ def project_point(self, point: array_like) -> Point:
...
+
+        Parameters
+        ----------
+        points : array_like
+            Input points.
+
+        Returns
+        -------
+        Points
+            Projection of the points onto the plane.
```
Comment (ajhynes7) [inline src/skspatial/objects/plane.py:288]:
```suggestion
            Projections of the points onto the plane.
```

File: src/skspatial/objects/plane.py [hunk 8]
```diff
@@ -392,6 +423,38 @@ def distance_point_signed(self, point: array_like) -> np.float64:
 
         return self.normal.scalar_projection(vector_to_point)
 
+    def distance_points_signed(self, points: array_like) -> np.ndarray:
+        """
+        Return the signed distances from points to the plane.
```
Comment (ajhynes7) [inline src/skspatial/objects/plane.py:428]:
```suggestion
        Return the signed distances from multiple points to the plane.
```

File: src/skspatial/objects/plane.py [hunk 9]
```diff
@@ -428,6 +491,36 @@ def distance_point(self, point: array_like) -> np.float64:
         """
         return abs(self.distance_point_signed(point))
 
+    def distance_points(self, point: array_like) -> np.ndarray:
+        """
+        Return the distances from points to the plane.
```
Comment (ajhynes7) [inline src/skspatial/objects/plane.py:496]:
```suggestion
        Return the distances from multiple points to the plane.
```

File: tests/unit/objects/test_line.py [hunk 10]
```diff
@@ -145,6 +146,27 @@ def test_project_point(point, point_line, vector_line, point_expected, dist_expe
...
+def test_project_points(points, point_line, vector_line):
+    line = Line(point_line, vector_line)
+
+    distances = line.distance_points(points)
+    points_projected = line.project_points(points)
+
+    distanecs_expected = [line.distance_point(point) for point in points]
+    points_expected = [line.project_point(point) for point in points]
+
+    assert np.all(np.isclose(points_projected, points_expected))
```
Comment (ajhynes7) [inline tests/unit/objects/test_line.py:166]: I think `np.allclose` would be more appropriate to use here.

**Review by ajhynes7 — CHANGES_REQUESTED** [review]
Hi @EdYazbec,

I agree with Cristiano that the code and docstrings maintain the style of the project, so thanks for paying attention to that!

I have a few suggested edits for the text, and some other comments.

**Comment by EdYazbec** [comment]
Thanks for your feedback! I'll get these implemented and go from there.

File: src/skspatial/objects/plane.py [hunk 11]
```diff
@@ -428,6 +491,36 @@ def distance_point(self, point: array_like) -> np.float64:
...
         return abs(self.distance_point_signed(point))
 
+    def distance_points(self, point: array_like) -> np.ndarray:
+        """
+        Return the distances from multiple points to the plane.
+
+        Parameters
+        ----------
+        points : array_like
+            Input point.
```
Comment (ajhynes7) [inline src/skspatial/objects/plane.py:501]:
```suggestion
            Input points.
```

File: tests/unit/objects/test_line.py [hunk 12]
```diff
@@ -145,6 +146,27 @@ def test_project_point(point, point_line, vector_line, point_expected, dist_expe
...
+        ([[0, 5], [1, 5], [0, 1], [1, 0], [0, 2], [-15, 5], [50, 10]], [1, -5], [0, 3])
+    ],
+)
+def test_project_points(points, point_line, vector_line):
+    line = Line(point_line, vector_line)
+
+    distances = line.distance_points(points)
+    points_projected = line.project_points(points)
+
+    distanecs_expected = [line.distance_point(point) for point in points]
```
Comment (ajhynes7) [inline tests/unit/objects/test_line.py:163]:
```suggestion
    distances_expected = [line.distance_point(point) for point in points]
```

File: tests/unit/objects/test_line.py [hunk 13]
```diff
@@ -192,6 +214,23 @@ def test_distance_point(array_point, line, dist_expected):
...
+        ([[0, 0], [8, 7], [20, -3]], Line([0, 0], [1, 0])),
+        ([[0, 0], [8, 7], [20, -3]], Line([0, 0], [0, 1])),
+        ([[0, 0], [8, 7], [20, -3]], Line([0, 1], [1, 0])),
+        ([[0, 0, 0], [8, 7, 13], [20, -3, 1]], Line([0, 0, 0], [1, 0, 0]))
+    ],
+)
+def test_distance_points(array_points, line):
+    distances = line.distance_points(array_points)
+
+    distanecs_expected = [line.distance_point(point) for point in array_points]
```
Comment (ajhynes7) [inline tests/unit/objects/test_line.py:229]:
```suggestion
    distances_expected = [line.distance_point(point) for point in array_points]
```

File: tests/unit/objects/test_plane.py [hunk 14]
```diff
@@ -140,6 +141,31 @@ def test_project_point(point, point_plane, normal_plane, point_expected, dist_ex
...
+    ],
+)
+def test_project_points(points, point_plane, normal_plane):
+    plane = Plane(point_plane, normal_plane)
+
+    distances = plane.distance_points(points)
+    points_projected = plane.project_points(points)
+    distances_signed = plane.distance_points_signed(points)
+
+    distances_expected = [plane.distance_points(point) for point in points]
```
Comment (ajhynes7) [inline tests/unit/objects/test_plane.py:160]:
```suggestion
    distances_expected = [plane.distance_point(point) for point in points]
```

File: tests/unit/objects/test_plane.py [hunk 15]
```diff
@@ -240,6 +266,26 @@ def test_distance_point(point, plane, dist_signed_expected):
...
+        ([[0, 0, 0], [50, -67, 0], [50, -67, 0], [5, 3, 8], [5, 3, 7], [5, 3, -8]], Plane([0, 0, 0], [0, 0, 1])),
+        ([[0, 0, 0], [50, -67, 0], [50, -67, 0], [5, 3, 8], [5, 3, 7], [5, 3, -8]], Plane([0, 0, 1], [0, 0, 1])),
+        ([[0, 0, 0], [50, -67, 0], [50, -67, 0], [5, 3, 8], [5, 3, 7], [5, 3, -8]], Plane([0, 0, 0], [0, 0, -50]))
+    ],
+)
+def test_distance_points(points, plane):
+    distances = plane.distance_points(points)
+    distances_signed = plane.distance_points_signed(points)
+
+    distanecs_expected = [plane.distance_point(point) for point in points]
```
Comment (ajhynes7) [inline tests/unit/objects/test_plane.py:281]:
```suggestion
    distances_expected = [plane.distance_point(point) for point in points]
```

File: tests/unit/objects/test_plane.py [hunk 16]
```diff
@@ -240,6 +266,26 @@ def test_distance_point(point, plane, dist_signed_expected):
...
+    distances_signed_expected = [plane.distance_points_signed(point) for point in points]
```
Comment (ajhynes7) [inline tests/unit/objects/test_plane.py:282]:
```suggestion
    distances_signed_expected = [plane.distance_point_signed(point) for point in points]
```

File: tests/unit/objects/test_plane.py [hunk 17]
```diff
@@ -240,6 +266,26 @@ def test_distance_point(point, plane, dist_signed_expected):
...
+    
+    assert np.allclose(distances_signed, distances_signed_expected)
+    assert np.allclose(distances, np.abs(distances_signed_expected))
+    assert np.allclose(distances, distanecs_expected)
```
Comment (ajhynes7) [inline tests/unit/objects/test_plane.py:286]:
```suggestion
    assert np.allclose(distances, distances_expected)
```

**Review by ajhynes7 — CHANGES_REQUESTED** [review]

File: tests/unit/objects/test_line.py [hunk 18]
```diff
@@ -192,6 +214,23 @@ def test_distance_point(array_point, line, dist_expected):
...
+        ([[0, 0], [8, 7], [20, -3]], Line([0, 1], [1, 0])),
+        ([[0, 0, 0], [8, 7, 13], [20, -3, 1]], Line([0, 0, 0], [1, 0, 0]))
+    ],
+)
+def test_distance_points(array_points, line):
+    distances = line.distance_points(array_points)
+
+    distances_expected = [line.distance_point(point) for point in array_points]
+
+    assert np.allclose(distances, distanecs_expected)
```
Comment (ajhynes7) [inline tests/unit/objects/test_line.py:231]:
```suggestion
    assert np.allclose(distances, distances_expected)
```

File: tests/unit/objects/test_line.py [hunk 19]
```diff
@@ -145,6 +146,27 @@ def test_project_point(point, point_line, vector_line, point_expected, dist_expe
...
+    line = Line(point_line, vector_line)
+
+    distances = line.distance_points(points)
+    points_projected = line.project_points(points)
+
+    distances_expected = [line.distance_point(point) for point in points]
+    points_expected = [line.project_point(point) for point in points]
+
+    assert np.allclose(points_projected, points_expected)
+    assert np.allclose(distances, distanecs_expected)
```
Comment (ajhynes7) [inline tests/unit/objects/test_line.py:167]:
```suggestion
    assert np.allclose(distances, distances_expected)
```

**Comment by ajhynes7** [comment]
Looks good! Thanks for contributing these. I'll get a release out soon.

**Pull request merged**